### PR TITLE
Fix File-related analysis plugin hooks

### PR DESCRIPTION
**Pull request by lhchavez**
This change does a few things:

* Renames `BeforeAnalyzeFileInterface` to `BeforeFileAnalysisInterface`.
* Adds a few more parameters to `BeforeFileAnalysisInterface`.
* Since the analysis was never called due to a bug, the above two
  changes are safe to do.
* Fix the above-mentioned bug, and now adds the hooks to the config.
* Update the documentation about the File-related analysis plugin hooks.
* Add a test to ensure that this does not break again in the future.

Fixes: #3969

### CODE REVIEW

File: docs/running_psalm/plugins/authoring_plugins.md [hunk 1]
```diff
@@ -68,11 +68,12 @@ class SomePlugin implements \Psalm\Plugin\Hook\AfterStatementAnalysisInterface
 - `AfterCodebasePopulatedInterface` - called after Psalm has scanned necessary files and populated codebase data.
 - `AfterEveryFunctionCallAnalysisInterface` - called after Psalm evaluates any function call. Cannot influence the call further.
 - `AfterExpressionAnalysisInterface` - called after Psalm evaluates an expression.
+- `AfterFileAnalyisisInterface` - called after Psalm analyzes a file.
```
Comment (andrew-demb) [inline docs/running_psalm/plugins/authoring_plugins.md:71]: Typo

Comment (lhchavez) [reply docs/running_psalm/plugins/authoring_plugins.md:71]: Fixed!

**Comment by muglug** [comment]
Thanks!

**Comment by TysonAndre** [comment]
> Renames `BeforeAnalyzeFileInterface` to `BeforeFileAnalysisInterface`.

I'd consider this to be a backwards incompatible API change requiring a new major version (e.g. a third party plugin using that would break).
Instead, it might make more sense to have `BeforeAnalyzeFileInterface` extend from `BeforeFileAnalysisInterface` and be `@deprecated use BeforeFileAnalysisInterface instead in 3.x.y`

**Comment by weirdan** [comment]
@TysonAndre, according to @lhchavez it never worked in the first place.

**Comment by TysonAndre** [comment]
In that case, never mind, no objections.

I guess I'd fixed this at some point but forgot to make a PR to contribute that upstream after validating that bug fixes/merge conflict fixes worked.

**Pull request merged**